### DISCUSSION TITLE
Realization of Sigma protocol

### DISCUSSION THREAD

**Discussion by Davidyzy** [jedisct1/libsodium#1088] [Q&A] [answered]
I am trying to use libsodium to realize a three-round sigma protocol. There is one part need calculation of g^r*(g^e)^w=g^(r+ew). However, I cannot get the equation that (g^e)^w=g^(e*w). I am afraid that it is due to the scalar multiplication and the mod L in the function. Can anyone help me with this? Thank you. The below gew and gwe should be the same but in fact they are different.

    // g^w
    unsigned char w[crypto_core_ed25519_SCALARBYTES];
    unsigned char gw[crypto_core_ed25519_BYTES];
    crypto_core_ed25519_scalar_random(w);
    crypto_scalarmult_ed25519_base_noclamp(gw, w);
    // h^e=(g^w)^e
    unsigned char e[crypto_core_ed25519_SCALARBYTES];
    unsigned char gwe[crypto_core_ed25519_BYTES];
    crypto_core_ed25519_scalar_random(e);
    crypto_scalarmult_ed25519(gwe, e, gw);
    unsigned char ew[crypto_core_ed25519_SCALARBYTES];
    //e*w
    crypto_core_ed25519_scalar_mul(ew,e,w);
    unsigned char gew[crypto_core_ed25519_BYTES];
    //g^(e*w)
    crypto_scalarmult_ed25519_base_noclamp(gew, ew);

**Comment by jedisct1** [comment, marked as answer]
> However, I cannot get the equation that (g^e)^w=g^(e*w)

```c
crypto_scalarmult_ed25519_base_noclamp(ge, e);    // g^e
crypto_core_ed25519_scalar_mul(ew, e, w);         // ew
crypto_scalarmult_ed25519_base_noclamp(g_ew, ew); // g^ew
crypto_scalarmult_ed25519_noclamp(ge_w, w, ge);   // ge^w
```

→ `g_ew` ≡ `ge_w`
